File: kunlun-analysis/src/main/java/com/hansight/kunlun/analysis/statistics/spark/HistoryAnomalyDetection.scala
package com.hansight.kunlun.analysis.statistics.spark

import java.math.BigDecimal
import java.net.InetAddress
import java.util
import com.hansight.kunlun.analysis.realtime.model.RTConstants
import com.hansight.kunlun.analysis.realtime.single.StringEntropyCalculator
import com.hansight.kunlun.analysis.statistics.model.{TimeIPPair, TimeTotalMaps}
import com.hansight.kunlun.analysis.utils.quantile.QDigest
import com.hansight.kunlun.analysis.utils.{DatetimeUtils, EventTypeUtils}
import org.apache.hadoop.io.Writable
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}
import org.elasticsearch.hadoop.cfg.ConfigurationOptions
import org.elasticsearch.hadoop.mr.LinkedMapWritable
import org.elasticsearch.spark._

import scala.collection.mutable
import scala.collection.JavaConverters._

/**
 * Author: zhhui 
 * Date: 2014/9/16
 */
object HistoryAnomalyDetection extends Serializable {
  var modelUrlRate: Double = .0
  var modelSc404Rate: Double = .0
  var modelSc500Rate: Double = .0

  def main(args: Array[String]) {
    if (args.length < 2) throw new IllegalArgumentException("args is less ,must be 2 args")
    val input: Array[String] = args(1).split("/")
    if (input.length < 2) {
      throw new IllegalArgumentException("args is less,must be ../..")
    }
    val master = args(0)
    val readIndex = input(0)
    val readType = input(1)
    val writeIndex = "anomaly"
    val writeType = "anomaly"
    compute(master, readIndex, readType, writeIndex, writeType)
  }


  def compute(master: String,
              readIndex: String,
              readType: String,
              writeIndex: String,
              writeType: String): Unit = {
    val conf = new SparkConf().setAppName("Simple Application").setMaster("local[12]")
    conf.set(ConfigurationOptions.ES_NODES, master + ":9200").set("es.scroll.size", "5000")
    val sc = new SparkContext(conf)
    val modelData = sc.esRDD("model_final/anomaly")
    val model = modelData.first()._2

    val urlRate = model.get("urlRate")
    val sc404 = model.get("sc404")
    val sc500 = model.get("sc500")
    if (urlRate != None)
      HistoryAnomalyDetection.modelUrlRate = urlRate.get.asInstanceOf[Double]
    if (sc404 != None)
      HistoryAnomalyDetection.modelSc404Rate = sc404.get.asInstanceOf[Double]
    if (sc500 != None)
      HistoryAnomalyDetection.modelSc500Rate = sc500.get.asInstanceOf[Double]
    val logData = sc.esRDD(readIndex + "/" + readType).asInstanceOf[RDD[Map[String, Any]]]
    val logs = logData.filter(filterNone).map(mapToTuple).groupByKey().map(reduce)
    logs.saveToEs(writeIndex + "/" + writeType)
  }

  /**
   * Output Key: TimeIPPair, combination of [date + time + c_ip]<br/>
   * Output Value: LinkedMapWritable of [input]<br/>
   */
  def filterNone(value: Map[String, Any]): Boolean = {
    val ip = value.get("c_ip")
    val dt = value.get("datetime")
    val uri_query = value.get("cs_uri_query")
    if (ip == None || dt == None || uri_query == None)
      true
    false
  }

  def mapToTuple(kv: Map[String, Any]): (TimeIPPair, mutable.Map[String, Writable]) = {
    val value = kv.values.head.asInstanceOf[mutable.Map[String, Writable]]
    val ip = value.get("c_ip")
    val c_ip: String = if (ip == None) "" else ip.get.toString
    val dt = value.get("datetime")
    val datetime: String = if (dt == None) "0" else dt.get.toString
    val uri_query = value.get("cs_uri_query")
    val query: String = if (uri_query == None) "" else uri_query.get.toString
    val qvs = new LinkedMapWritable()
    // val entropy: Double = StringEntropyCalculator.calculate(query)
    query.split('&').foreach(pair => {
      val posEQ: Int = pair.indexOf('=')
      val (param, entropy) = if (posEQ >= 0) (pair.substring(0, posEQ), StringEntropyCalculator.calculate(pair.substring(posEQ + 1))) else ("", StringEntropyCalculator.calculate(pair))
      qvs.put(param, entropy)
    })
    value += (("cs_uri_query", qvs))
    val timestamp: Long = DatetimeUtils.getTimestamp(datetime)
    val ipaddr: InetAddress = InetAddress.getByName(c_ip)
    var clientIP: Long = 0
    for (b <- ipaddr.getAddress) clientIP = clientIP << 8 | (b & 0xFF)
    (new TimeIPPair(timestamp, clientIP), value)
  }

  /**
   * <pre>
   * Input &lt;K,Iterator&lt;V&gt;&gt; pair as following:<br/>
   * &lt;TimeIPPair, Iterator&lt;"cs-uri-stem:x,cs-uri-query:x,cs_referer:x,cs-host:x,sc-status:x"&gt;&gt;
   */
  def reduce(kvs: (TimeIPPair, Iterable[mutable.Map[String, Writable]])): TraversableOnce[Map[String, Any]] = {
    val key2: TimeIPPair = new TimeIPPair(kvs._1)
    val timeTotalMaps: TimeTotalMaps = new TimeTotalMaps(key2.getTime)
    val servers: util.Set[String] = new util.HashSet[String]
    kvs._2.foreach(log => {
      val serverValue = log.get("s_computer_name")
      if (serverValue != None)
        servers.add(serverValue.get.toString)
      val uriStemValue = log.get("cs_uri_stem")
      val hostValue = log.get("cs_host")
      if (uriStemValue != None) {
        val k: String = hostValue.get.toString + "/" + uriStemValue.get.toString
        val value:Integer = timeTotalMaps.getUniqueUrls.asScala.get(k).getOrElse(0)
        timeTotalMaps.getUniqueUrls.put(k, value+ 1)
      }
      val statusValue = log.get("sc_status")
      if (statusValue != None) {
        val k: Integer = statusValue.get.asInstanceOf
        val value:Integer = timeTotalMaps.getResponseCodes.asScala.get(k).getOrElse(0)
        timeTotalMaps.getResponseCodes.put(k, value+ 1)
      }
      val qvs = log.get("cs_uri_query")
      if (qvs != None) {
      }
      timeTotalMaps.inc()

    })
    val endDate = kvs._2.last.get("time")

    val urlRate = if (timeTotalMaps.getUniqueUrls.size > 0) {
      val digest: QDigest = new QDigest(timeTotalMaps.getUniqueUrls.size + 1)
      import scala.collection.JavaConversions._
      timeTotalMaps.getUniqueUrls.values.foreach(num => digest.offer(num.longValue))
      new BigDecimal(digest.getQuantile(0.1).toDouble / timeTotalMaps.getTotal.toDouble).setScale(5, BigDecimal.ROUND_CEILING).doubleValue
    } else {
      0.0
    }
    val smaps = timeTotalMaps.getResponseCodes.asScala
    val sc404=smaps.get(404)
    val sc404Rate: Double = if (sc404 != None) new BigDecimal(sc404.get.toDouble / timeTotalMaps.getTotal).setScale(5, BigDecimal.ROUND_CEILING).doubleValue else 0.0
    val sc500=smaps.get(500)
    val sc500Rate: Double = if (sc500!= None) new BigDecimal(sc500.get.toDouble / timeTotalMaps.getTotal).setScale(5, BigDecimal.ROUND_CEILING).doubleValue else 0.0
    var logs: List[Map[String, Any]] = Nil
    val map = Map("c_ip" -> key2.getIp,
      "endDatetime" -> endDate.toString,
      "startDatetime" -> key2.getTime,
      "servers" -> servers.toArray(new Array[String](servers.size)))
    if (urlRate > HistoryAnomalyDetection.modelUrlRate) {

      logs = map +(("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP_UNIQUE_URL)),
        ("eventType", RTConstants.EVENT_TYPE_HTTP_UNIQUE_URL),
        ("degree", urlRate),
        ("modelDegree", HistoryAnomalyDetection.modelUrlRate)) :: logs
    }
    if (sc404Rate > HistoryAnomalyDetection.modelSc404Rate) {
      logs = map +(("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP404)),
        ("eventType", RTConstants.EVENT_TYPE_HTTP404),
        ("degree", sc404Rate),
        ("modelDegree", HistoryAnomalyDetection.modelUrlRate)) :: logs
    }
    if (sc500Rate > HistoryAnomalyDetection.modelSc500Rate) {
      logs = map +(("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP500)),
        ("eventType", RTConstants.EVENT_TYPE_HTTP500),
        ("degree", sc500Rate),
        ("modelDegree", HistoryAnomalyDetection.modelUrlRate)) :: logs
    }
    logs
  }
}

File: kunlun-analysis/src/main/java/com/hansight/kunlun/analysis/statistics/mr/HistoryAnomalyDetection.java
package com.hansight.kunlun.analysis.statistics.mr;

import com.hansight.kunlun.analysis.realtime.model.RTConstants;
import com.hansight.kunlun.analysis.realtime.single.StringEntropyCalculator;
import com.hansight.kunlun.analysis.statistics.model.AnomalyConstants;
import com.hansight.kunlun.analysis.statistics.model.TimeIPPair;
import com.hansight.kunlun.analysis.statistics.model.TimeTotalMaps;
import com.hansight.kunlun.analysis.utils.DatetimeUtils;
import com.hansight.kunlun.analysis.utils.EventTypeUtils;
import com.hansight.kunlun.analysis.utils.quantile.QDigest;
import com.hansight.kunlun.utils.EsUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.io.*;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Partitioner;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.log4j.Logger;
import org.codehaus.jettison.json.JSONException;
import org.codehaus.jettison.json.JSONObject;
import org.elasticsearch.action.search.SearchResponse;
import org.elasticsearch.client.transport.TransportClient;
import org.elasticsearch.hadoop.cfg.ConfigurationOptions;
import org.elasticsearch.hadoop.mr.EsInputFormat;
import org.elasticsearch.hadoop.mr.EsOutputFormat;
import org.elasticsearch.hadoop.mr.LinkedMapWritable;
import org.elasticsearch.index.query.QueryBuilders;
import org.elasticsearch.search.SearchHit;

import java.io.IOException;
import java.math.BigDecimal;
import java.net.InetAddress;
import java.util.HashSet;
import java.util.Set;

/**
 * c_Ip：客户IP地址，cs_referer：访问到的地址 ，sc_status：返回状态码
 */
public class HistoryAnomalyDetection extends Configured implements Tool {
	private static final Logger LOG = Logger
			.getLogger(HistoryAnomalyDetection.class);
	public static String esMaster;
	public static String esReadIndex;
	public static String esReadType;
	public static String esWriteIndex;
	public static String esWriteType;

	/**
	 * The first field of TimeIPPair is the start time, the second is the
	 * client,请求到来的第一次时间，IP地址 IP
	 */
	public static class SessionMapper extends
			Mapper<Text, LinkedMapWritable, TimeIPPair, LinkedMapWritable> {
		static final Logger s_logger = Logger.getLogger(SessionMapper.class);
		protected String[] fields;

		/**
		 * Output Key: TimeIPPair, combination of [date + time + c_ip]<br/>
		 * Output Value: LinkedMapWritable of [input]<br/>
		 */
		@Override
		protected void map(Text key, LinkedMapWritable value, Context context)
				throws IOException, InterruptedException {
			try {
				String c_ip = value.get(AnomalyConstants.IP).toString();
				String datetime = value.get(AnomalyConstants.dtKey).toString();
				String query = value.get(AnomalyConstants.uriQuery).toString();
				MapWritable qvs = new LinkedMapWritable();
				// Split into param-value pairs
				double entropy = StringEntropyCalculator.calculate(query);

				int posNameStart = 0; // Move from the start pos of query string
										// to the first name
				do {
					int posValueEnd = query.indexOf('&', posNameStart);
					if (posValueEnd < 0)
						posValueEnd = query.length();

					String pair = query.substring(posNameStart, posValueEnd);
					int posEQ = pair.indexOf('=');

					String param = null, pvalue = null;
					if (posEQ >= 0) {
						param = pair.substring(0, posEQ);
						pvalue = pair.substring(posEQ + 1);
					} else {
						param = "";
						pvalue = pair;
					}
					qvs.put(new Text(param), new DoubleWritable(entropy));
					posNameStart = posValueEnd + 1;
				} while (posNameStart < query.length());
				value.put(AnomalyConstants.uriQuery, qvs);
				// ignore '.000Z' to avoid parse exception, time unit is 1 sec
				// at least
				long timestamp = DatetimeUtils.getTimestamp(datetime);
				InetAddress ipaddr = InetAddress.getByName(c_ip);
				long clientIP = 0;
				for (byte b : ipaddr.getAddress())
					clientIP = clientIP << 8 | (b & 0xFF);
				context.write(new TimeIPPair(timestamp, clientIP), value);
			} catch (Exception ex) {
				s_logger.error(ex.getMessage());
			} finally {
				// logDigest.clear();
			}
		}

	}

	public static class SessionReducer extends
			Reducer<TimeIPPair, LinkedMapWritable, NullWritable, Text> {
		private static final Integer ONE = 1;
		static final Logger s_logger = Logger.getLogger(SessionReducer.class);

		private double modelUrlRate;
		private double modelSc404Rate;
		private Double modelSc500Rate;
		private String readIndex;



		@Override
		protected void setup(Context context) throws IOException,
				InterruptedException {
			Configuration conf = context.getConfiguration();
			modelUrlRate = conf.getDouble("anomaly.urlRate", 0);
			modelUrlRate = conf.getDouble("anomaly.sc404", 0);
			modelUrlRate = conf.getDouble("anomaly.sc500", 0);
			readIndex = conf.get(ConfigurationOptions.ES_RESOURCE_READ);
		}

		/**
		 * <pre>
		 * Input &lt;K,Iterator&lt;V&gt;&gt; pair as following:<br/>
		 * &lt;TimeIPPair, Iterator&lt;"cs-uri-stem:x,cs-uri-query:x,cs_referer:x,cs-host:x,sc-status:x"&gt;&gt;
		 */
		@Override
		protected void reduce(TimeIPPair key,
				Iterable<LinkedMapWritable> values, Context context)
				throws IOException, InterruptedException {
			TimeIPPair key2 = new TimeIPPair(key);
			s_logger.info("Key info: " + key2.toString());
			TimeTotalMaps timeTotalMaps = new TimeTotalMaps(key.getTime());
			LinkedMapWritable lastLog = null;
			Set<String> servers = new HashSet<>();
			for (LinkedMapWritable log : values) {// log 相同IP
				lastLog = log;
				Writable serverValue = log.get(AnomalyConstants.computername);
				servers.add(serverValue.toString());
				
				Writable uriStemValue = log.get(AnomalyConstants.uriStem);
				Writable hostValue = log.get(AnomalyConstants.host);
				if (uriStemValue != null) {
					String k = hostValue.toString() + "/"
							+ uriStemValue.toString();// c_ip一段时间内请求一个地址的次数
					Integer value = timeTotalMaps.getUniqueUrls().get(k);
					timeTotalMaps.getUniqueUrls().put(k,
							value == null ? ONE : value + 1);
				}
				Writable statusValue = log.get(AnomalyConstants.status);
				if (statusValue != null) {
					Integer k = Integer.parseInt(statusValue.toString());
					Integer value = timeTotalMaps.getResponseCodes().get(k);
					timeTotalMaps.getResponseCodes().put(k,
							value == null ? ONE : value + 1);
				}
				Writable qvs = log.get(AnomalyConstants.uriQuery);
				if (qvs != null) {
					// TODO: the query parameter mode here

				}
				timeTotalMaps.inc();
			}
			Writable endDate = lastLog.get(AnomalyConstants.time);
			double urlRate = 0.0, sc404Rate = 0.0, sc500Rate = 0.0;
			if (timeTotalMaps.getUniqueUrls().size() > 0) {
				QDigest digest = new QDigest(timeTotalMaps.getUniqueUrls()
						.size() + 1);
				for (Integer num : timeTotalMaps.getUniqueUrls().values()) {
					s_logger.info("UniqueUrls: " + num);
					digest.offer(num.longValue());
				}
				long l = digest.getQuantile(0.1);
				urlRate = new BigDecimal((double) l / timeTotalMaps.getTotal())
						.setScale(5, BigDecimal.ROUND_CEILING).doubleValue();

			}
			if (timeTotalMaps.getResponseCodes().get(404) != null)
				sc404Rate = new BigDecimal((double) timeTotalMaps
						.getResponseCodes().get(404) / timeTotalMaps.getTotal())
						.setScale(5, BigDecimal.ROUND_CEILING).doubleValue();
			if (timeTotalMaps.getResponseCodes().get(500) != null)
				sc500Rate = new BigDecimal((double) timeTotalMaps
						.getResponseCodes().get(500) / timeTotalMaps.getTotal())
						.setScale(5, BigDecimal.ROUND_CEILING).doubleValue();
			try {
                JSONObject json = new JSONObject();
				json.put("c_ip", key.getIp());
				json.put("endDatetime", endDate.toString());
				json.put("startDatetime", key.getTime());
				json.put("indices", readIndex);
				json.put("servers", servers.toArray(new String[servers.size()]));
				if (urlRate > modelUrlRate) {
					json.put("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP_UNIQUE_URL));
					json.put("eventType", RTConstants.EVENT_TYPE_HTTP_UNIQUE_URL);
					json.put("degree", urlRate);
					json.put("modelDegree", modelUrlRate);
					context.write(NullWritable.get(), new Text(json.toString()));
				}
				if (sc404Rate > modelSc404Rate) {
					json.put("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP404));
					json.put("eventType", RTConstants.EVENT_TYPE_HTTP404);
					json.put("degree", sc404Rate);
					json.put("modelDegree", modelSc404Rate);
					context.write(NullWritable.get(), new Text(json.toString()));
				}
				if (sc500Rate > modelSc500Rate) {
					json.put("category", EventTypeUtils.getCategory(RTConstants.EVENT_TYPE_HTTP500));
					json.put("eventType", RTConstants.EVENT_TYPE_HTTP500);
					json.put("degree", sc500Rate);
					json.put("modelDegree", modelSc500Rate);
					context.write(NullWritable.get(), new Text(json.toString()));
				}
			} catch (JSONException e) {
				s_logger.error(e.getMessage());
			}

			// Get the session unique URL quantile (10 percentile), 404/500 rate
			// for (Map.Entry<Long, TimeTotalMaps> entry : map.entrySet()) {}

		}
	}
	
	public static class FirstPartitioner extends Partitioner<TimeIPPair, Text> {
		// static final Logger s_logger =
		// Logger.getLogger(FirstPartitioner.class);
		@Override
		public int getPartition(TimeIPPair key, Text value, int numPartitions) {
			// s_logger.info("key: " + key + ", value: " + value.toString() +
			// ", partitions: "+ numPartitions);
			// divide by IP
			return (int) (Math.abs(key.getIp() * 127) % numPartitions);
		}
	}

	public static class KeyComparator extends WritableComparator {
		// static final Logger s_logger = Logger.getLogger(KeyComparator.class);
		protected KeyComparator() {
			super(TimeIPPair.class, true);
		}

		/**
		 * Sort will follow TimeIPPair comparator sequence
		 */
		@Override
		public int compare(WritableComparable w1, WritableComparable w2) {
			TimeIPPair p1 = (TimeIPPair) w1;
			TimeIPPair p2 = (TimeIPPair) w2;
			// s_logger.info("compare: " + p1 + " and " + p2);
			int cmp = TimeIPPair.compare(p1.getIp(), p2.getIp());
			if (cmp != 0) {
				return cmp;
			}
			return TimeIPPair.compare(p1.getTime(), p2.getTime());
		}
	}

	public static class GroupComparator extends WritableComparator {
		// static final Logger s_logger =
		// Logger.getLogger(GroupComparator.class);
		protected GroupComparator() {
			super(TimeIPPair.class, true);
		}

		/**
		 * Logs with from the same c-ip within 30 minutes will go to the same
		 * reducer<br/>
		 */
		@Override
		public int compare(WritableComparable w1, WritableComparable w2) {
			TimeIPPair p1 = (TimeIPPair) w1;
			TimeIPPair p2 = (TimeIPPair) w2;
			// s_logger.info("compare: " + p1 + " and " + p2);

			int cmp = TimeIPPair.compare(p1.getIp(), p2.getIp());
			if (cmp != 0) {
				return cmp;
			}
			// return cmp;
			/* 取消分组 */
			long timeDiff = p1.getTime() - p2.getTime();
			if (timeDiff > 300000) // 1800000L
				return 1; // as one session if there are within 30 mins
			else if (timeDiff < -300000)
				return -1;
			else
				return 0;

		}
	}

	@Override
	public int run(String[] args) throws Exception {
		if (args.length != 1) {
			usage();
			return 1;
		}
		String date = args[0];
		esReadIndex = "logs_" + date;
		esReadType = "aadp";
		esWriteIndex = "anomaly_" + date;
		esWriteType = "anomaly";
		Configuration conf = new Configuration();
		conf.addResource("es-site.xml");
		conf.set(ConfigurationOptions.ES_RESOURCE_READ,
				String.format("%s/%s/", esReadIndex, esReadType));
		conf.set(ConfigurationOptions.ES_RESOURCE_WRITE,
				String.format("%s/%s/", esWriteIndex, esWriteType));
		if (!setModelRate(conf, date)) {
			throw new IllegalArgumentException(
					"unable get the model from model_final/anomaly");
		}

		// ES model_final/anomaly
		Job job = Job.getInstance(conf);
		job.setJarByClass(HistoryAnomalyDetection.class);

		job.setInputFormatClass(EsInputFormat.class);
		job.setOutputFormatClass(EsOutputFormat.class);

		job.setMapperClass(SessionMapper.class);
		job.setMapOutputKeyClass(TimeIPPair.class);
		job.setMapOutputValueClass(LinkedMapWritable.class);

		job.setPartitionerClass(FirstPartitioner.class);
		job.setSortComparatorClass(KeyComparator.class);
		job.setGroupingComparatorClass(GroupComparator.class);

		job.setReducerClass(SessionReducer.class);
		job.setOutputKeyClass(NullWritable.class);
		job.setOutputValueClass(Text.class);
		return job.waitForCompletion(true) ? 0 : 1;
	}

	private boolean setModelRate(Configuration conf, String date) {
		String name = conf.get("es.cluster.name");
		String address = conf.get("es.nodes");
		String index = "model_final";
		String type = "anomaly";
		String field = "date";
		TransportClient client = EsUtils.getNewClient(name, address);
		SearchResponse response = client.prepareSearch(index).setTypes(type)
				.setQuery(QueryBuilders.termQuery(field, date)).setFrom(0)
				.setSize(1).execute().actionGet();
		if (response != null) {
			for (SearchHit hit : response.getHits()) {
				Double urlRate = (Double) hit.getSource().get("urlRate");
				Double sc404 = (Double) hit.getSource().get("sc404");
				Double sc500 = (Double) hit.getSource().get("urlRate");
				conf.set("anomaly.urlRate", urlRate.toString());
				conf.set("anomaly.sc404", sc404.toString());
				conf.set("anomaly.sc500", sc500.toString());
				return true;
			}
		}
		return false;
	}

	private void usage() {
		System.out.println("params: date");
		System.out.println("date: yyyyMMdd, example: 20131215");
	}

	public static void main(String[] args) throws Exception {
		int exitCode = ToolRunner.run(new AnomalyDetection(), args);
		System.exit(exitCode);
	}
}